src/homeblks/volume/volume.cpp

```diff
@@ -335,7 +335,105 @@ indx_tbl* Volume::recover_indx_tbl(btree_super_block& sb, btree_cp_sb& cp_info)
     return static_cast< indx_tbl* >(tbl);
 }
 
+static std::vector< std::pair< int, int > > get_true_intervals(const std::vector< bool >& empty_blocks) {
+    std::vector< std::pair< int, int > > result;
+
+    int start = -1;
+    for (std::size_t i = 0; i < empty_blocks.size(); ++i) {
+        if (empty_blocks[i]) {
+            if (start == -1) { start = i; }
+        } else {
+            if (start != -1) {
+                result.emplace_back(start, i - start);
+                start = -1;
+            }
+        }
+    }
+    if (start != -1) { result.emplace_back(start, empty_blocks.size() - start); }
+    return result;
+}
+
+static std::vector< std::pair< int, int > > compute_range_intervals(const uint8_t* buf, size_t page_size,
+                                                                    uint32_t nlbas, bool empty_blocks = false) {
+    std::vector< std::pair< int, int > > intervals;
+    bool in_empty_region = false;
+    int current_range_start = -1;
+    int current_range_length = 1;
+    for (uint32_t i = 0; i < nlbas; i++) {
+        const uint8_t* page_start = buf + (i * page_size);
+        bool is_page_empty = (empty_blocks == is_buf_zero(page_start, page_size));
+        if (is_page_empty) {
+            if (!in_empty_region) {
+                current_range_start = i;
+                current_range_length = 1;
+                in_empty_region = true;
+            } else {
+                current_range_length++;
+            }
+        } else {
+            if (in_empty_region) { intervals.push_back(std::make_pair(current_range_start, current_range_length)); }
+            in_empty_region = false;
+        }
+    }
+    if (in_empty_region) { intervals.push_back(std::make_pair(current_range_start, current_range_length)); }
+    return intervals;
+}
+
+static bool is_buf_zero(const uint8_t* buf, size_t size) {
+    static std::vector< uint8_t > read_buf(size, 0);
+    static const auto zero_crc = crc16_t10dif(init_crc_16, read_buf.data(), size);
+    const auto crc = crc16_t10dif(init_crc_16, buf, size);
+    return (crc == zero_crc) ? (buf[0] == 0 && !std::memcmp(buf, buf + 1, size - 1)) : false;
+}
+
+static bool is_iovec_zero(const std::vector< iovec >& iovecs) {
+    for (const auto& iovec : iovecs) {
+        auto data = static_cast< uint8_t* >(iovec.iov_base);
+        const size_t size = iovec.iov_len;
+        if (!is_buf_zero(data, size)) { return false; }
+    }
+    return true;
+}
+static bool is_zero_request(const vol_interface_req_ptr& iface_req, uint32_t page_size) {
+    if (iface_req->iovecs.empty()) {
+        return is_buf_zero(static_cast< uint8_t* >(iface_req->buffer), iface_req->nlbas * page_size);
+    }
+    return is_iovec_zero(iface_req->iovecs);
+}
+
+static std::string print_ranges(lba_t start_lba, const std::vector< std::pair< int, int > >& intervals) {
```

Comment:
this one also can be removed if not going to be used

Reply:
Can you put these three APIs that are currently not used together and make an #if 0 around them, so that it would be easier for others to read and understand this thin-provisioning code? Also put comment that these apis are for future usage if we support paritial zero detection.

Reply:
sure